Integrated the new `NavigationServer` and `NavigationServer2D`

AndreaCatania opened the pull request:
This is the integration of the `NavigationServer` proposed here: https://github.com/godotengine/godot-proposals/issues/332

# What's new
#### Retro compatibility
This new `NavigationServer`, that is used for both 3D and 2D part, even if it offers many new features it's fully retro compatible.

#### Map & Regions welding
The `Navigation` node now creates a new `Map` in the server, and is possible to add as many `Region`s you desire by simply adding `NavigationMeshInstance` under the `Navigation` node (perfectly equal as before). The difference is that the `Regions` now are welded all together to form one single fully navigable map.
This operation is cheap because it doesn't regenerate the `NavigationMesh`, rather it just connects the many (probably pre-baked) `Regions`. You can remove and add `Regions` as you like at runtime.

### `NavigationMesh` baking
The `NavigationMesh` can be pre baked in the editor (as before), or can be baked during the game without stopping it. This operation is not cheap and for this reason it's done in a separate thread, once it's done, it automatically sets the new `NavigationMesh` without any stop. (See the below 3D gif)

### `NavigationAgent`
Even if the old workflow is still valid, now it's possible to use the `NavigationAgent` node that simplifies a lot the map navigation.
The `NavigationAgent` just needs to know the desired final location; it internally takes care to compute the optimal path. This path may or may not change during the game, but what you have to only care about is to just reach the next location returned by: `$NavigationAgent.get_next_location()`, let the agent do the work for you.

### Collision Avoidance
Since the `NavigationMesh` baking is an heavy operation, an algorithm for collision avoidance that take cares of `Dynamic` obstacles got introduced.
As mention before, you can compute the player velocity using the position returned by the `get_next_location()` function but this function doesn't take care of the dynamic obstacles.
To avoid collisions you have to set the calculated velocity to the agent, using: `$NavigationAgent.set_velocity(velocity)`, then the event: `_velocity_computed(safe_velocity):` is emitted with the parameter `safe_velocity` that is a version of your velocity vector that avoids dynamic obstacles.

### Obstacles
The collision avoidance takes care to avoid collision with other agents and obstacles. Agents and Obstacles are different only for the fact that Obstacles can't be directly controlled. Similar to the Agents, is possible to create an Obstacle by adding the node `NavigationObstacle`.
The `NavigationObstacle` doesn't have any parameter because it's automatically initialized and updated, so you have just to add it to the scene.

### 2D
The `Navigation2DServer` is a bridge to the `NavigationServer` that allow to use the latter using 2D classes, like `Vector2D` or `Transform2D`.
The nodes `NavigationAgent2D` and `NavigationObstacle2D` works exactly like the 3D version described above.

### Multi thread safe
The `NavigationServer` keeps tracks of any call and executes these during the __sync phase__. This mean that you can request any change to your map using any thread without any worrying.

### 3D version
In the below _GIF_ you can see a character that navigates the map, composed by two different regions, while avoiding collisions with dynamic objects. Most important, you can see that after a bit the right side region is modified and a call to `bake` allows me to `Regenerate` the navigation mesh and so reach the above part before not reachable.

Also note that the pink box is a kinematic object, while the black balls are rigid bodies.

![ezgif com-video-to-gif(3)](https://user-images.githubusercontent.com/8342599/71680159-aa59c200-2d89-11ea-9b54-669e94c49319.gif)

In this _GIF_ it's possible to see that moving a region, automatically executes the map re-link and immediately allows the navigation. This step is done in real time.

![ezgif com-video-to-gif(4)](https://user-images.githubusercontent.com/8342599/71713455-21369f80-2e0a-11ea-8043-8c701134a403.gif)


### 2D version
Similarly to the above 3D version, the character is navigating two regions while avoiding the dynamic obstacles collisions.

![ezgif com-video-to-gif(2)](https://user-images.githubusercontent.com/8342599/71680161-aa59c200-2d89-11ea-84d9-4056a399938f.gif)

### Example project
[Navigation2.zip](https://github.com/godotengine/godot/files/4021788/Navigation2.zip)

_This work has been kindly sponsored by IMVU._

Duroxxigar commented:
I wholeheartedly support this @AndreaCatania. As it is something that has been a pain point for Godot for many years. I do not see anything done for the docs side however. It would be quite helpful to have that part documented as much as possible, doc writers can come in and pretty it up if need be. However, it would be better to have at least a foundation in place before they roll in.

(I don't mean to distract from the wonderful job here - just looking at it from a community stand point).

reduz commented:
Awesome work! Will throughly review the coming week

Duroxxigar commented:
@AndreaCatania I noticed in the previous PR that you had, you did document everything in the .xml files, but I'm not seeing it in this PR. Perhaps I overlooked it?

Zylann commented on modules/gdnavigation/navigation_mesh_generator.cpp:
```diff
@@ -53,15 +60,15 @@
+void NavigationMeshGenerator::_add_vertex(const Vector3 &p_vec3, Vector<float> &p_verticies) {
 	p_verticies.push_back(p_vec3.x);
 	p_verticies.push_back(p_vec3.y);
 	p_verticies.push_back(p_vec3.z);
```
You may want to resize + 3 first and set the 3 last values instead, `Vector` calls `realloc` on every `push_back`

AndreaCatania replied:
This is a problem of the Vector that doesn't have a proper a reserve mechanism. I prefer to not take into account locally, so when it will be improved I can benefit from it automatically.

Zylann commented on scene/resources/ray_shape.cpp:
```diff
@@ -41,6 +41,10 @@ Vector<Vector3> RayShape::get_debug_mesh_lines() {
 }
 
+real_t RayShape::get_enclosing_radius() const {
+	return length;
```
Not really a radius? `length` would be twice as big. In some places it returns a proper radius, sometimes it doesnt.

AndreaCatania replied:
The ray shape is casted always from the origin of the shape, so a sphere / circle that encloses it must have the same radius of the ray length.

However, the final radius of a shape depends also on it's `Shape` node transformation relative its parent.
So if the Ray length is 1 and it's offsetted to x0, y-3, z0 it's the shape total length is 3.

Zylann commented on scene/resources/cylinder_shape.cpp:
```diff
@@ -62,6 +62,10 @@ Vector<Vector3> CylinderShape::get_debug_mesh_lines() {
 }
 
+real_t CylinderShape::get_enclosing_radius() const {
+	return Vector2(radius, height).length();
```
For the radius of an enclosing sphere, shouldn't it be `Vector2(radius, height*0.5).length()`?

AndreaCatania replied:
Yep, fixed

avencherus commented:
Looking forward to this.  Though I didn't make it too far in testing yet, the engine crashes.  (Win10 64-bit 118aaa55f9577c9)

In the 2D (Level3) example I copied the Rigidbody2Ds, since the interesting thing to me about RVO would be how lots of entities navigate and with different sizes and speeds.

![image](https://user-images.githubusercontent.com/13004169/71763337-91fdba80-2ee3-11ea-9a82-06a7ad80140f.png)



>	godot.windows.tools.64.exe!std::_Default_allocator_traits<std::allocator<RVO::Plane> >::construct<RVO::Plane,RVO::Plane>(std::allocator<RVO::Plane> & __formal, RVO::Plane * const _Ptr, RVO::Plane && <_Args_0>) Line 761	C++
 	godot.windows.tools.64.exe!std::_Uninitialized_backout_al<std::allocator<RVO::Plane> >::_Emplace_back<RVO::Plane>(RVO::Plane && <_Vals_0>) Line 1600	C++
 	godot.windows.tools.64.exe!std::_Uninitialized_move<RVO::Plane *,std::allocator<RVO::Plane> >(RVO::Plane * const _First, RVO::Plane * const _Last, RVO::Plane * _Dest, std::allocator<RVO::Plane> & _Al) Line 1699	C++
 	godot.windows.tools.64.exe!std::vector<RVO::Plane,std::allocator<RVO::Plane> >::_Umove(RVO::Plane * _First, RVO::Plane * _Last, RVO::Plane * _Dest) Line 1559	C++
 	godot.windows.tools.64.exe!std::vector<RVO::Plane,std::allocator<RVO::Plane> >::_Emplace_reallocate<RVO::Plane const &>(RVO::Plane * const _Whereptr, const RVO::Plane & <_Val_0>) Line 709	C++
 	godot.windows.tools.64.exe!std::vector<RVO::Plane,std::allocator<RVO::Plane> >::emplace_back<RVO::Plane const &>(const RVO::Plane & <_Val_0>) Line 661	C++
 	godot.windows.tools.64.exe!std::vector<RVO::Plane,std::allocator<RVO::Plane> >::push_back(const RVO::Plane & _Val) Line 671	C++
 	godot.windows.tools.64.exe!RVO::Agent::computeNewVelocity(float timeStep) Line 189	C++
 	godot.windows.tools.64.exe!NavMap::compute_single_step(unsigned int _index, RvoAgent * * agent) Line 577	C++
 	godot.windows.tools.64.exe!ThreadArrayProcessData<NavMap,RvoAgent * *>::process(unsigned int p_index) Line 50	C++
 	godot.windows.tools.64.exe!process_array_thread<ThreadArrayProcessData<NavMap,RvoAgent * *> >(void * ud) Line 64	C++
 	godot.windows.tools.64.exe!ThreadWindows::thread_callback(void * userdata) Line 55	C++
 	[External Code]	


![image](https://user-images.githubusercontent.com/13004169/71763313-4f3be280-2ee3-11ea-830d-e19333408cc9.png)

AndreaCatania commented:
@avencherus I've fixed a problem, I'm quite sure it was related. Can you please check it now? If it will continue to not work I'll test it in windows.

However, I've updated the sample project by adding more characters and other changes to allow it:

[Navigation2.zip](https://github.com/godotengine/godot/files/4021787/Navigation2.zip)

![2020-01-04 11-16-31](https://user-images.githubusercontent.com/8342599/71764321-d2ab0300-2ee6-11ea-96d3-25ed1b6fb076.gif)

AndreaCatania commented:
> @AndreaCatania I noticed in the previous PR that you had, you did document everything in the .xml files, but I'm not seeing it in this PR. Perhaps I overlooked it?

@Duroxxigar No, you are correct. That PR was based on the RVO only feature and that docs was written by @nathanmerrill; that documentation is no more valid, for this reason you can't find it here.

However, if someone want to document it, I'm available to help him.

balenol commented:
AT LAST!
Now I don't need to implement my own collision avoidance system in 2D by clipping the navigation polygon. Thanks bro.

Zireael07 commented:
I'm more happy with the merge navigation meshes/nav polygons part - I was trying to achieve merging in GDScript and had to ultimately scrap that in favor of using A* for pathing.

Duroxxigar commented:
@AndreaCatania Good to know. I've been eagerly looking forward to something like this for quite some time. Once I got some downtime this weekend, I was going to start working on the documentation myself. Most likely starting this evening (so in about 5-6 hours hopefully). I want this thing to be well understood.

starry-abyss commented:
Great work!
Does it by any chance support multiple layers of navigation maps? For example, one for ground agents, one for swimming agents, and one for flying agents.

aaronfranke commented on servers/navigation_server.h:
```diff
@@ -0,0 +1,192 @@
+///
+/// Note: All the `set` functions are commands executed during the `sync` phase,
+/// don't expect that a change is immediately propagated.
+class NavigationServer : public Object {
```
Since this is a change slated for Godot 4.0, and 3D nodes will be renamed to have 3D suffixes, shouldn't this addition be called `NavigationServer3D`?

AndreaCatania replied:
I saw that, but this is not yet a decision made nor we are in a transition state. So I thought that it's not yet optimal a name like: `NavigationServer3D`.
We could change idea in any moment and use another kind of naming strategy.

avencherus commented:
@AndreaCatania Thanks.  Gave it a retest this morning.  The changes appear to have resolved the crash, and your example project works well.  Though the agents are a bit jittery and sometimes become sluggish, is this expected, and to be resolved by the end implementation of the game?

Also I reran the project where I fiddle about and created the crash, and oddly the agents ignore the walls in many cases.

![godot windows tools 64_2020-01-05_10-23-42](https://user-images.githubusercontent.com/13004169/71777265-2cbecd80-2fa6-11ea-946d-f140effb3cc3.png)
![godot windows tools 64_2020-01-05_10-23-45](https://user-images.githubusercontent.com/13004169/71777266-30eaeb00-2fa6-11ea-850f-3ab7638d5d45.png)

After some time, a few agents at the end of goal will restart their pathing.

Here is my project so you can inspect what I did.  (This is from Level3)

[RVO.zip](https://github.com/godotengine/godot/files/4023010/RVO.zip)

Naryosha commented on servers/navigation_2d_server.cpp:
```diff
@@ -0,0 +1,237 @@
+	ClassDB::bind_method(D_METHOD("agent_set_radius", "agent", "radius"), &Navigation2DServer::agent_set_radius);
+	ClassDB::bind_method(D_METHOD("agent_set_max_speed", "agent", "max_speed"), &Navigation2DServer::agent_set_max_speed);
+	ClassDB::bind_method(D_METHOD("agent_set_velocity", "agent", "velocity"), &Navigation2DServer::agent_set_velocity);
+	ClassDB::bind_method(D_METHOD("agent_set_velocity_target", "agent", "target_velocity"), &Navigation2DServer::agent_set_target_velocity);
```
I think it should be either `target_velocity` or `velocity_target` for consistency. Same for the 3D variant. I like the latter more because it'd show up next to `agent_set_velocity` in the docs (although I'm not sure there is any naming conventions in use.)

starry-abyss replied:
I think "velocity target" is the least confusing. "Target velocity" sounds similar to "target's velocity" (velocity of some other guy besides the agent), while "velocity target" is more obviously the level of velocity to be reached.

aaronfranke replied:
Also, shouldn't it be `set_agent_radius` instead of `agent_set_radius`? I think there's a discussion about this specifically somewhere, and I can't find it, but https://github.com/godotengine/godot/issues/33026 is related.

Zylann replied:
When I see such naming I think of the convention used in servers in general, as they are close to a C-like API, so because they use to take a RID as "self", the type of "self" would be prepended as the first word of the function's name

AndreaCatania commented:
> Great work!
> Does it by any chance support multiple layers of navigation maps? For example, one for ground agents, one for swimming agents, and one for flying agents.

It's already possible. Each `Navigation` node is a map. You can have three maps one for the _Walking_ area one for _Swimming_ area, the last one for _Flying_ area.

AndreaCatania commented:
@Naryosha Updated

AndreaCatania commented:
@avencherus 
> Though the agents are a bit jittery and sometimes become sluggish, is this expected, and to be resolved by the end implementation of the game?

Yes this is expected. You have to set the agent properties (like `Time Horizon`, `Max Speed`) properly so the Agent can decide to react earlier or later to an obstacle.

> Also I reran the project where I fiddle about and created the crash, and oddly the agents ignore the walls in many cases.

This is expected since the walls doesn't have any collider.

> After some time, a few agents at the end of goal will restart their pathing.

This is a feature. The `Agent` doesn't control the `start` nor the `end` of the path resolution, but this is up to the developer decide how to use it. In my case, I've decided to run the path resolution endless so even if the character reached the final destination it can still fight to be more near possible to the end.

Duroxxigar commented on scene/3d/navigation_mesh_instance.cpp:
```diff
@@ -0,0 +1,258 @@
+	ADD_PROPERTY(PropertyInfo(Variant::BOOL, "enabled"), "set_enabled", "is_enabled");
+
+	ADD_SIGNAL(MethodInfo("navigation_mesh_changed"));
+	ADD_SIGNAL(MethodInfo("bake_done"));
```
Should this signal be named `bake_finished`? I believe other signals throughout the engine have it in this way. I don't recall a signal with "done" for the name. (Only suggesting for constancy sake)

AndreaCatania replied:
There are other signals with the `_done` suffix, like:
- `_thumbnail_done`
- `_tree_thumbnail_done`
- `_file_list_thumbnail_done`
- `_preview_done`
- `_thread_done`
- `_request_done`

I didn't found any event with `_finish` suffix.
Despite this, do you think it would make more sense: `bake_finish`?

Calinou replied:
@AndreaCatania We have signals that end with `completed` and `finished`, so we should use past tense, not present tense.

Duroxxigar replied:
@AndreaCatania The first one that pops in my mind is with animations. I believe audio is "finished" as well. I wasn't aware of the other signals you listed. Other than the `_request_done` one, but that one slipped my mind.

@Calinou If this is the case - does this call in the question some of the ones that they mentioned?

AndreaCatania replied:
My fault, in the fast check that I did this morning I searched for: `_finish"` instead of `_finished"`.
I've already changed it, it's `bake_finished` now.

AndreaCatania commented:
Quick update, I've improved the path finding and added some other events.

- When the target is near a unreachable polygon:
__Now:__ The `Agent` returns a path that bring the player most near possible to the target.
__Before:__ The `get_simple_path` returns a void path.

- When the `Agent` path is reloaded:
__Now:__ The `Agent` checks for the best next sub location.
__Before:__ The `Agent` resets its navigation to the beginning.

# Before
![71713455-21369f80-2e0a-11ea-8043-8c701134a403](https://user-images.githubusercontent.com/8342599/71993929-33647380-3238-11ea-8c8d-2cd86459fe4e.gif)


# Now
As you can see now the Player is able to navigate toward the target even if the navigation mesh changes, and its target location is unreachable.
![ezgif com-video-to-gif(5)](https://user-images.githubusercontent.com/8342599/71993889-23e52a80-3238-11ea-9597-ebf18fbce7df.gif)

starry-abyss commented:
@AndreaCatania Just to make sure. What I was interested in is:
- flying is possible in special areas + in ground areas + in swimming areas;
- swimming agents can only use swimming areas;
- amphibious agents can use both swimming and ground areas.

Like in Warcraft 3, for example.
With multiple areas suggestion it's possible to reconnect them on the fly, but is it possible to have them connected for some agents and at the same time not connected for other agents?

AndreaCatania committed:
- Integrated NavigationServer and Navigation2DServer.
- Added Navigation Agents and Obstacles.
- Integrated Collision Avoidance.

This work has been kindly sponsored by IMVU.

AndreaCatania commented:
@starry-abyss Path finding for Flying and Swimming characters using a navigation mesh is not the right technology and I want to advice use AStar os something similar.

The reason is that 3D Navigation Mesh doesn't sound optimal, however RVO would work really well so if you have small obstacles you can consider using a pure RVO approach.

> - swimming agents can only use swimming areas;
> - amphibious agents can use both swimming and ground areas.

I'm pretty sure it's already doable with a bit of work.

Zylann commented:
@starry-abyss Depending on your flying model you could consider using layers of navmeshes, and your flyer could switch between those layers.

starry-abyss commented:
@Zylann Flying agents are not like in first/third-person action game, but like in RTS. They always are at certain height from the ground (no need to even switch layers!), and layers are a good option I think. But how is it achieved in Godot baking (how to say Godot to bake different navmeshes from the same level?), and does Godot further support agent reading from different layers (and not just hardcoded to always have one)?
Sorry, it may not be fully related to the PR, but to Godot in general, no matter old or future version.

HeadClot commented:
Got a question for you @AndreaCatania - Does this PR support [nav mesh off links](https://docs.unity3d.com/Manual/class-OffMeshLink.html)? Just curious as this is a really useful feature for jumping and Climbing AI. :)

slapin commented:
Any ledge detecttion support?
Also did somebody compared this t more mature solution like recastnavigation libs feature-wise?

AndreaCatania commented:
@HeadClot This is an interesting feature, which is not so difficult to integrate either. I'm not sure that I can do it in the near future but completely doable.

@slapin Ledge detection is something that depends a lot on how the character motion works and a generic integration of this feature is not really so useful. Let's suppose that your character can teleport 2 meters away or can jump here and there... A generic ledge detection is not so useful.
But, what's your idea?

slapin commented:
The polygon marking thing like in Detour is very useful to limit/filter
navmesh queries when finding closest ledge.
You can easily define what ledge is using agent parameters. See how Detour
works. The feature is very useful
for motion controllers beyond primitive and it is very hard/ineffective to
do it differently. So that is usually
implemented on top of generic navmesh marking.
Also I do not understand what is your concern with generic ledge detection
and why it is not useful.
Ledges are all polygons on side of navmesh flat region which are walkable
but are adjacent to edge of navmesh or
unwalkable area. So they are quite generic. The detection of unwalkable
areas depends on parameters of course,
but these are still have to be calculated anyway. See the queries which
Detour supports, that might give you some
ideas what is generally expected of such a system.

On Sat, Jan 25, 2020 at 11:49 AM Andrea Catania <notifications@github.com>
wrote:

> @HeadClot <https://github.com/HeadClot> This is an interesting feature,
> which is not so difficult to integrate either. I'm not sure that I can do
> it in the near future but completely doable.
>
> @slapin <https://github.com/slapin> Ledge detection is something that
> depends a lot on how the character motion works and a generic integration
> of this feature is not really so useful. Let's suppose that your character
> can teleport 2 meters away or can jump here and there... A generic ledge
> detection is not so useful.
> But, what's your idea?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/34776?email_source=notifications&email_token=AAABPUYICQMFJHKPXFOM4L3Q7P4J5A5CNFSM4KCK57Y2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJ4YIII#issuecomment-578389025>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAABPU7YJFBUMQ3ZHOPU7D3Q7P4J5ANCNFSM4KCK57YQ>
> .
>

folays commented:
Please consider going forward and try to this merge this.

I think I read previously, that some times ago @slapin did some attempt of providing good work in this area (navigation), which did not receive attention, and disappointed him, and made him abandon it.

Now @AndreaCatania seems to make an attempt again, of providing a sound implementation of recastnavigation.

Maybe it is now perfect but it could be a step forward to a real progress.

I think lot of people have an experience in the general recastnavigation API, and thus once this "base work" in this proposal is merged, more people could rejoin in adding anything missing, because lot of people already known more-or-less how the API should looks, and what it should allow to do.

Some people maybe need a leap of faith. recastnavigation has medium documentation, lots of blog post on it, and heavy usage which helps to figure how to use it.

I don't think integrating this merge early would cause a delayed/missing release window.

AndreaCatania commented:
@slapin I was thinking about ledge detection being done using the physics engine (through queries)... But you are right, a function to return ledges would be not only easy to add, but also useful!

reduz commented:
Some comments so far:

* Is there any point in having a NavigationServer2D? is it not easier and simpler to just make the 2D nodes use the regular NavigationServer
* If we are using Detour, does it not make more sense to just call it NavigationServerDetour ?
* Servers should probably not have getters unless its used to get bulk data (like collision, meshes), do you think it makes sense to remove them?
* I don't quite understand the command system, should we not just use the MT thread wrapper used in the physics and visual server so it can kind of be called from multiple threads?

reduz commented:
Ok after talking to Odino this sounds like it should be ok to merge, but i guess up to you if you want to do it after/before rebasing master

akien-mga commented on thirdparty/README.md:
```diff
@@ -439,6 +439,20 @@ Files extracted from upstream source:
+- README
+
+Important: Some files have Godot-made changes; so to enrich the features originally proposed by this library and better integrate this library with Godot. Please check the file to know what's new.
+
```
There should be two lines between each section, so one missing here.

AndreaCatania replied:
Done

akien-mga commented on thirdparty/README.md:
```diff
@@ -439,6 +439,20 @@ Files extracted from upstream source:
+## Rvo2-3D
+
+- Upstream: http://gamma.cs.unc.edu/RVO2/
+- version: 1.0.1
```
Uppercase "Version".

akien-mga replied:
I just saw that it's wrong for `recastnavigation` above too, so if you want to fix it at the same time.

AndreaCatania replied:
Done

akien-mga commented on thirdparty/README.md:
```diff
@@ -439,6 +439,20 @@ Files extracted from upstream source:
 - License.txt
 
 
+## Rvo2-3D
```
Use the same name as the folder, so `rvo2`.

AndreaCatania replied:
Done

akien-mga commented on thirdparty/README.md:
```diff
@@ -439,6 +439,20 @@ Files extracted from upstream source:
+
+- All .cpp and .h files in the `src/` folder except for RVOSimulator.cpp and RVOSimulator.h
+- LICENSE
+- README
```
No need to include the `README`.

AndreaCatania replied:
Done

akien-mga commented on thirdparty/README.md:
```diff
@@ -439,6 +439,20 @@ Files extracted from upstream source:
+- LICENSE
+- README
+
+Important: Some files have Godot-made changes; so to enrich the features originally proposed by this library and better integrate this library with Godot. Please check the file to know what's new.
```
Please wrap the line to <=80 chars.

Are Godot-made changes to the library well identified?

AndreaCatania replied:
yep

AndreaCatania replied:
Done

akien-mga commented on doc/classes/NavigationAgent.xml:
```diff
@@ -0,0 +1,163 @@
+		</signal>
+		<signal name="target_reached">
+			<description>
+				Notifies when the player defined target, set with [method set_target], is reached.
```
Should be `[method set_target_location]`.

AndreaCatania replied:
Corrected

Duroxxigar replied:
Ahh well...that is embarrassing, lol.

akien-mga commented on doc/classes/NavigationAgent2D.xml:
```diff
@@ -0,0 +1,157 @@
+		</signal>
+		<signal name="target_reached">
+			<description>
+				Notifies when the player defined target, set with [method set_target], is reached.
```
Same here.

AndreaCatania replied:
Corrected

Duroxxigar committed:
Documented the new NavigationServer and all its associated nodes (2D and 3D)

akien-mga commented:
Thanks!

slapin commented:
Please avoid meaningless comments when doing review. Keep them clean and to
the point. Even when under influence.

On Mon, Feb 10, 2020 at 7:24 PM Duroxxigar <notifications@github.com> wrote:

> *@Duroxxigar* commented on this pull request.
> ------------------------------
>
> In doc/classes/NavigationAgent.xml
> <https://github.com/godotengine/godot/pull/34776#discussion_r377169817>:
>
> > +		</member>
> +	</members>
> +	<signals>
> +		<signal name="navigation_finished">
> +			<description>
> +				Notifies when the final location is reached.
> +			</description>
> +		</signal>
> +		<signal name="path_changed">
> +			<description>
> +				Notifies when the navigation path changes.
> +			</description>
> +		</signal>
> +		<signal name="target_reached">
> +			<description>
> +				Notifies when the player defined target, set with [method set_target], is reached.
>
> Ahh well...that is embarrassing, lol.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/godotengine/godot/pull/34776?email_source=notifications&email_token=AAABPU3YKBMWI6N5FQX4BFLRCF5VRA5CNFSM4KCK57Y2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCU4UU6I#discussion_r377169817>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAABPUZQ6NCPTSYWCSEJEUTRCF5VRANCNFSM4KCK57YQ>
> .
>

akien-mga commented:
> Please avoid meaningless comments when doing review. Keep them clean and to
the point. Even when under influence.

Please avoid meaningless comments about comments you find meaningless... Especially not referring to a fantasized consumption of substance.

mourning20s commented on scene/3d/navigation.h:
```diff
@@ -31,154 +31,45 @@
-	Vector3 get_closest_point_to_segment(const Vector3 &p_from, const Vector3 &p_to, const bool &p_use_collision = false);
-	Vector3 get_closest_point(const Vector3 &p_point);
-	Vector3 get_closest_point_normal(const Vector3 &p_point);
-	Object *get_closest_point_owner(const Vector3 &p_point);
```
Were these functions intended to be removed? It'll be pretty difficult to write 3D point-and-click navigation code without these.

AndreaCatania replied:
I may add these back on the agent; before this can you please explain the use case, so I know better how it are intended to be used?

mourning20s replied:
Personally, I use `get_closest_point_to_segment` to raycast from a camera (e.g. so a player can click where they want a character to go). This functionality is also present in the 3D navmesh demo if you'd like a working example: https://github.com/godotengine/godot-demo-projects/tree/master/3d/navmesh

AndreaCatania replied:
The idea is to use the physics server to get this info, for this reason I removed these, but add this feature may be useful. I'll add it, thanks.

Zireael07 replied:
I think I was using that, too - converting 2D map to 3D vectors, and then finding which segment (road) I was closest to (when creating the map, so in a tool script, where physics is not available nor necessary)

mourning20s replied:
@AndreaCatania Thank you for taking the time to do that

slapin-omp replied:
But how do you intend to use physics server for querying navmesh nformation?

AndreaCatania replied:
Fixed by this #36269

TheSHEEEP commented:
Is there any plan for this to be backported to 3.2 in one of the LTS releases?

I'm asking since I'm almost done with version 0.1 of my own navigation module (focused on procedural generation, so no editor integration) for Godot 3.2, and now I am wondering if I should continue working on it beyond that.

https://github.com/TheSHEEEP/godotdetour

Though from what I've seen so far, both projects have very different foci and features. Eg. godotdetour supports multiple navmeshes at the same time from the get-go, as well as marking areas as water, grass, etc. and I might add other things as I need them.
And maybe most importantly, godotdetour uses detour (duh!) for everything, while NavigationServer uses RVO2 instead (not sure why, to be honest).

aaronfranke commented:
@TheSHEEEP This would be very unlikely to be backported to 3.2, since it's a major change and has already broken compatibility ([here](https://github.com/godotengine/godot/pull/34776#discussion_r377279947)). Generally, for something to be backported it would either be a useful or important bug fix, or an important feature that's very unlikely to cause regressions.

TheSHEEEP commented:
I thought so, just better to ask than to continue working on it if there's a significant risk it might just get obsolete the moment it is "done" ;)

Though I might have continued anyway due to the different foci of the libraries.

bruvzg commented:
This PR breaks MinGW/clang cross build form macOS (I have not tested it on Linux or with gcc) with following error:

```
In file included from modules/gdnavigation/register_types.cpp:42:
In file included from modules/gdnavigation/navigation_mesh_editor_plugin.h:36:
In file included from ./editor/editor_node.h:34:
In file included from ./editor/editor_data.h:38:
./editor/plugins/script_editor_plugin.h:141:3: error: expected identifier
                FILE_OPEN,
                ^
.../llvm-mingw/x86_64-w64-mingw32/include/winnt.h:4900:43: note: expanded from macro 'FILE_OPEN'
#define FILE_OPEN                         0x00000001
                              
```

owstetra commented:
i get this Error when Compiling on Windows with MinGW :
```
In file included from ./scene/3d/navigation.h:34,\n                 from ./mod
ules/gridmap/grid_map.h:34,\n                 from modules\\recast\\navigation_m
esh_generator.cpp:54:\n./scene/3d/navigation_mesh_instance.h:40:7: error: redefi
nition of 'class NavigationMeshInstance'\n class NavigationMeshInstance : public
 Spatial {\n       ^~~~~~~~~~~~~~~~~~~~~~\nIn file included from modules\\recast
\\navigation_mesh_generator.h:35,\n                 from modules\\recast\\naviga
tion_mesh_generator.cpp:31:\n./scene/3d/navigation_mesh.h:198:7: note: previous
definition of 'class NavigationMeshInstance'\n class NavigationMeshInstance : pu
blic Spatial {\n       ^~~~~~~~~~~~~~~~~~~~~~\n"
=====
scons: *** [modules\recast\navigation_mesh_generator.windows.tools.64.o] Error 1

scons: building terminated because of errors.
```

Sorry , i copy it directly from CMD

akien-mga commented:
And from Linux with MinGW/GCC :)
```
Compiling ==> thirdparty/rvo2/src/Agent.cpp
In file included from thirdparty/rvo2/src/Agent.h:40,
                 from thirdparty/rvo2/src/Agent.cpp:33:
thirdparty/rvo2/src/API.h:42:10: fatal error: SDKDDKVer.h: No such file or directory
 #include <SDKDDKVer.h>
          ^~~~~~~~~~~~~
```

akien-mga commented:
Why does Rvo2 rely on Windows SDK headers in the first place? That's bogus and should be fixed, the library should be cross-platform.

bruvzg commented:
Nothing in the `API.h` seems to be required, it's probably only for using RVO2 as dynamic library, I have reduced it to `#define RVO_API` and it builds on both macOS and MinGW.

crotron commented:
Does this change solve https://github.com/godotengine/godot/issues/28235?
